ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl/State.hs

```diff
@@ -97,32 +100,55 @@ data OpenState blockId h = OpenState {
 type ModifyOpenState m blockId h =
   StateT (OpenState blockId h) (WithTempRegistry (OpenState blockId h) m)
 
+data AppendOrWrite = Append | Write
+
 -- | NOTE: This is safe in terms of throwing FsErrors.
 modifyOpenState
-  :: forall blockId m a. (HasCallStack, IOLike m)
-  => VolatileDBEnv m blockId
+  :: forall blockId m a. IOLike m
+  => AppendOrWrite
+  -> VolatileDBEnv m blockId
   -> (forall h. Eq h => HasFS m h -> ModifyOpenState m blockId h a)
   -> m a
-modifyOpenState VolatileDBEnv {hasFS = hasFS :: HasFS m h, varInternalState} modSt = do
+modifyOpenState appendOrWrite
+                VolatileDBEnv {hasFS = hasFS :: HasFS m h, varInternalState}
+                modSt = do
     wrapFsError $ modifyWithTempRegistry getSt putSt (modSt hasFS)
   where
-    -- TODO Is uninterruptibleMask_ absolutely necessary here?
+    -- NOTE: we can't use the bracketed variants, as that's incompatible with
+    -- 'modifyWithTempRegistry', which takes a function to put back the state,
+    -- as that must have succeeded before the resources are released from the
+    -- temporary registry.
+    (acquire, release) = case appendOrWrite of
+      Append ->
+        (atomically .  RAWLock.unsafeAcquireAppendAccess,
+         atomically .: RAWLock.unsafeReleaseAppendAccess)
+      Write  ->
+        (RAWLock.unsafeAcquireWriteAccess, RAWLock.unsafeReleaseWriteAccess)
+
     getSt :: m (OpenState blockId h)
-    getSt = uninterruptibleMask_ $ takeMVar varInternalState >>= \case
+    getSt = acquire varInternalState >>= \case
       DbOpen ost -> return ost
       DbClosed   -> do
-        putMVar varInternalState DbClosed
+        release varInternalState DbClosed
         throwM $ UserError ClosedDBError
 
     putSt :: OpenState blockId h -> ExitCase (OpenState blockId h) -> m ()
-    putSt ost ec = do
-        -- It is crucial to replace the MVar.
-        putMVar varInternalState st'
-        unless (dbIsOpen st') $ closeOpenHandles hasFS ost
+    putSt ost ec = case closeOrOst' of
+        -- We must close the VolatileDB
+        Nothing   -> do
+          -- We don't care about the current state, as we were appending or
+          -- writing, which means that the state couldn't have changed in the
+          -- background.
+          _mbCurState <-
+              RAWLock.poison varInternalState (const (UserError ClosedDBError))
```

Comment:
Embed the exception in `ClosedDBError` (`Maybe SomeException` or two different constructors).